content/blogposts/2022/ranking-in-a-week.md

```diff
@@ -0,0 +1,120 @@
+
+The output data that Spark gives us is also a series of CSV file in a second GCS bucket, each with the following form:
+
+```csv
+repo,path,rank
+github.com/sourcegraph/blog-post-example,internal/handler.go,9.657402401
+github.com/sourcegraph/log,logger.go,15.567945095
+github.com/sourcegraph/blog-post-example,internal/database.go,10.999427919
+github.com/sourcegraph/blog-post-example,internal/experiments.go,8.604131051
+github.com/sourcegraph/blog-post-example,internal/init.go,7.266468968
+github.com/sourcegraph/blog-post-example,cmd/app/main.go,0.703945210
+...
+```
+
+Each CSV file (of which there are hundreds of thousands) is an equal-sized shard of results. The ranking scores for a single repository may be spread out over multiple files, thus we have to do some clever aggregation on our end to aggregate all of a repo's score. Thankfully, we have a [bit of expertise](https://thenewstack.io/how-to-stop-autovacuum-from-sucking-up-your-performance/) and [a bag of tricks](https://docs.sourcegraph.com/dev/background-information/sql/batch_operations) that help us ensure we can consume and aggregate this data without significantly increasing resource requirements on our primary database.
+
+On October 30th, we [updated the conditions](https://github.com/sourcegraph/sourcegraph/pull/43650) that tells our indexing server when a repository has had a relevant update. Before, this method would only return repositories that we know have received a Git push to remote since the last indexing attempt for that repository. After, this method will also return repositories for which we have an updated ranking scores. This allows us to pack the search indexes such that documents in a repository are ordered by their ranking.
+
+Although we've covered the bulk of the data pipeline, [other](https://github.com/sourcegraph/sourcegraph/pull/43817) [improvements](https://github.com/sourcegraph/sourcegraph/pull/43943) have [occurred](https://github.com/sourcegraph/sourcegraph/pull/43944), and we continue to iterate on ranking heuristics at every level of the stack (graph construction, PageRank algorithm variants, index-time rank packing of index shards, query-time resorting of search results, etc).
+
+### Looking to the future
+
+In a little bit over a week, we built a data pipeline to bring incredible new value to Sourcegraph that will fundamentally improve the user experience for code search and code navigation. As our Head of Engineering notes, we will continue to make improvements to our search relevance in every dimension over the coming days, weeks, months, and years. We have a pile of precise data, and now we have a proven way to compute different properties over our global source graph at scale.
```

Comment:
@efritz Could thoughts on adding a note that users can see the results of this work now on Sourcegraph.com?

Reply:
```suggestion

Ranking has launched last week for searches on popular OSS repos on [https://sourcegraph.com](https://sourcegraph.com) - try it out today!
```

Any other information to add here?

Reply:
@efritz Looks like you reworked the conclusion a little bit so this is outdated. I took a look at the new one though - I think I changed my mind - I like that the outro is purely about the technical story and I don't want to ruin it by pointing people back to the product & come across marketing-y. They can see the product link at the very start of the article. What do you think?

Reply:
```suggestion
In a little bit over a week, we built a data pipeline to bring incredible new value to Sourcegraph that will fundamentally improve the user experience for code search and code navigation. As our Head of Engineering notes, we will continue to make improvements to our search relevance in every dimension over the coming days, weeks, months, and years. We have a pile of precise data, and now we have a proven way to compute different properties over our global source graph at scale.
```

Better?

Reply:
```In a little bit over a week, we built a data pipeline to bring incredible new value to Sourcegraph that will fundamentally improve the user experience for code search and code navigation. As our Head of Engineering notes, we will continue to make improvements to our search relevance in every dimension over the coming days, weeks, months, and years. One improvement I'm particularly excited to jump into is running PageRank over much larger data with higher granularity: individual _symbols_ instead of _documents_.

We have a pile of precise data, and now we have a proven way to compute different properties over our global source graph at scale.```

Is what's currently in the article. Any changes to suggest?